nodebuilder/pruner/pruner.go

```diff
@@ -0,0 +1,308 @@
+	"github.com/ipfs/go-datastore/query"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/celestiaorg/celestia-node/header"
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+)
+
+const (
+	dsPrefix = "/pruner/epoch/"
+)
+
+var (
+	// GC is triggered when the number of active epochs exceeds the target epoch count by this amount
+	gcRatio = 1.1
+)
+
+var log = logging.Logger("pruner")
+
+type StoragePruner struct {
+	cancel context.CancelFunc
+	cfg    Config
+
+	oldestEpoch atomic.Uint64
+
+	stripedLocks [256]sync.Mutex
+	activeEpochs map[uint64]struct{}
+
+	ds    datastore.Batching
+	store *eds.Store
+
+	metrics *metrics
+
+	done chan struct{}
+}
+
+func NewStoragePruner(ds datastore.Batching, store *eds.Store, cfg Config) *StoragePruner {
+	sp := &StoragePruner{
+		activeEpochs: make(map[uint64]struct{}),
+		ds:           namespace.Wrap(ds, datastore.NewKey(dsPrefix)),
+		store:        store,
+		cfg:          cfg,
+		done:         make(chan struct{}),
+	}
+
+	// set to max uint64 as sentinel before state is restored or first epoch is registered
+	sp.oldestEpoch.Store(^uint64(0))
+	return sp
+}
+
+func (sp *StoragePruner) Start(ctx context.Context) error {
+	err := sp.restoreState(ctx)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	sp.cancel = cancel
+
+	go sp.run(ctx)
+	return nil
+}
+
+func (sp *StoragePruner) Stop(_ context.Context) error {
+	sp.cancel()
+	<-sp.done
+	return nil
+}
+
+func (sp *StoragePruner) restoreState(ctx context.Context) error {
+	results, err := sp.ds.Query(ctx, query.Query{})
+	if err != nil {
+		return fmt.Errorf("failed to recover pruner state from datastore: %w", err)
+	}
+	for {
+		res, ok := results.NextSync()
+		if !ok {
+			break
+		}
+		epoch, err := strconv.ParseUint(res.Key[1:], 10, 64)
+		if err != nil {
+			return fmt.Errorf("failed to parse epoch from datastore: %w", err)
+		}
+		// we don't need to use locks here because no methods will be called until the callers have also started
+		sp.activeEpochs[epoch] = struct{}{}
+		oldest := sp.oldestEpoch.Load()
+		if epoch < oldest {
+			sp.oldestEpoch.CompareAndSwap(oldest, epoch)
+		}
+	}
+	log.Infow("restored state from datastore",
+		"oldestEpoch", sp.oldestEpoch.Load(),
+		"active epoch count", len(sp.activeEpochs),
+	)
+	return nil
+}
+
+func (sp *StoragePruner) Register(ctx context.Context, h *header.ExtendedHeader) error {
+	var datahashes []share.DataHash
+	var err error
+
+	if share.DataHash(h.DAH.Hash()).IsEmptyRoot() {
+		return nil
+	}
+
+	epoch := sp.calculateEpoch(h.Time())
+	lk := &sp.stripedLocks[epoch%256]
+	lk.Lock()
+	defer lk.Unlock()
+	log.Debugf("registering datahash %X to epoch %d", h.DAH.Hash(), epoch)
+	_, ok := sp.activeEpochs[epoch]
+	if ok { // epoch already registered, load existing datahashes from datastore
+		datahashes, err = sp.getDatahashesFromEpoch(ctx, epoch)
+		if err != nil {
+			return err
+		}
```

Comment:
can you format all of this a bit more nicely (code blocking etc)

and short-circuit statement can be first.